doc/man/gnunet-tracekit.1
.TH GNUNET-TRACEKIT "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet\-tracekit \-\- overlay network traceing
.SH SYNOPSIS
.B gnunet\-tracekit
[\fIOPTIONS\fR]
.SH DESCRIPTION
gnunet\-tracekit can be used to visualize the GNUnet overlay network.  This tool broadcasts a message into the network querying every peer about a list of other connected peers. The result is then printed to the screen.
.PP
gnunet\-tracekit is primarily useful for developers to determine the structure of the network.  End\-users will find little use for the tool itself (except to satisfy their curiosity) but should at the moment still enable this protocol to help the developers that trace the network.
.PP

.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
load config file (defaults: ~/.gnunet/gnunet.conf)

.TP
\fB\-D \fIDEPTH\fR, \fB\-\-depth=\fIDEPTH\fR
until which depth should the network be traced (how many hops)

.TP
\fB\-F \fIFORMAT\fR, \fB\-\-format=\fIFORMAT\fR
specifies the output format, 0 is human readable, 1 is dot, 2 is vcg

.TP
\fB\-h\fR, \fB\-\-help\fR
print help page

.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).

\fB\-P \fIPRIORITY\fR, \fB\-\-priority=\fIPRIORITY\fR
what should the priority of the probe be
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number

\fB\-W \fIWAIT\fR, \fB\-\-wait=\fIWAIT\fR
set how many seconds to wait for replies

.SH NOTES
gnunet\-tracekit may not be supported by all GNUnet peers.  Peers that do not support tracekit will still be discovered and will be marked in the output as non-responding peers.
.SH FILES
.TP
~/.gnunet/gnunet.conf
Users GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/tracekit/gnunet-tracekit.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2004, 2006, 2007 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/tracekit/gnunet-tracekit.c
 * @brief tool that sends a trace request and prints the received network topology
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_protocols.h"
#include "gnunet_util.h"
#include "gnunet_tracekit_lib.h"
#include "tracekit.h"

struct SeenRecord
{
  GNUNET_PeerIdentity src;
  GNUNET_PeerIdentity dst;
};

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_GC_Configuration *cfg;

static unsigned int priority = 0;

static unsigned int depth = 5;

static unsigned int format = 0;

static unsigned int delay = 300;

static struct SeenRecord *seen;

static unsigned int count;

static int
check_seen (const GNUNET_PeerIdentity * src, const GNUNET_PeerIdentity * dst)
{
  static GNUNET_PeerIdentity null_peer;
  unsigned int j;

  if (dst == NULL)
    dst = &null_peer;
  for (j = 0; j < count; j++)
    if ((0 == memcmp (src,
                      &seen[j].src,
                      sizeof (GNUNET_HashCode))) &&
        (0 == memcmp (dst, &seen[j].dst, sizeof (GNUNET_HashCode))))
      return GNUNET_YES;
  GNUNET_array_grow (seen, count, count + 1);
  seen[count - 1].src = *src;
  seen[count - 1].dst = *dst;
  return GNUNET_NO;
}

/**
 * Generate a human-readable report.
 *
 * @param reporter identity of the peer reporting a connection
 * @param link identity of another peer that the reporting peer
 *             is reported to be connected to, or NULL if the
 *             peer is reporting to have no connections at all
 * @return GNUNET_OK to continue data gathering,
 *         GNUNET_SYSERR to abort
 */
static int
human_readable (void *unused,
                const GNUNET_PeerIdentity * reporter,
                const GNUNET_PeerIdentity * link)
{
  GNUNET_EncName src;
  GNUNET_EncName dst;

  if (check_seen (reporter, link))
    return GNUNET_OK;

  GNUNET_hash_to_enc (&reporter->hashPubKey, &src);
  if (link != NULL)
    {
      GNUNET_hash_to_enc (&link->hashPubKey, &dst);
      fprintf (stdout,
               _("`%s' connected to `%s'.\n"),
               (const char *) &src, (const char *) &dst);
    }
  else
    {
      fprintf (stdout,
               _("`%s' is not connected to any peer.\n"),
               (const char *) &src);
    }
  return GNUNET_OK;
}

/**
 * Generate dot-format.
 *
 * @param reporter identity of the peer reporting a connection
 * @param link identity of another peer that the reporting peer
 *             is reported to be connected to, or NULL if the
 *             peer is reporting to have no connections at all
 * @return GNUNET_OK to continue data gathering,
 *         GNUNET_SYSERR to abort
 */
static int
dot_format (void *unused,
            const GNUNET_PeerIdentity * reporter,
            const GNUNET_PeerIdentity * link)
{
  GNUNET_EncName src;
  GNUNET_EncName dst;

  if (check_seen (reporter, link))
    return GNUNET_OK;
  GNUNET_hash_to_enc (&reporter->hashPubKey, &src);
  if (link != NULL)
    {
      GNUNET_hash_to_enc (&link->hashPubKey, &dst);
      printf ("  \"%.*s\" -> \"%.*s\";\n",
              4, (char *) &src, 4, (char *) &dst);
    }
  else
    {
      printf ("  %.*s;\n", 4, (char *) &src);
    }

  return GNUNET_OK;
}

/**
 * Generate vcg-format.
 *
 * @param reporter identity of the peer reporting a connection
 * @param link identity of another peer that the reporting peer
 *             is reported to be connected to, or NULL if the
 *             peer is reporting to have no connections at all
 * @return GNUNET_OK to continue data gathering,
 *         GNUNET_SYSERR to abort
 */
static int
vcg_format (void *unused,
            const GNUNET_PeerIdentity * reporter,
            const GNUNET_PeerIdentity * link)
{
  GNUNET_EncName src;
  GNUNET_EncName dst;

  if (check_seen (reporter, link))
    return GNUNET_OK;
  GNUNET_hash_to_enc (&reporter->hashPubKey, &src);
  if (link != NULL)
    {
      GNUNET_hash_to_enc (&link->hashPubKey, &dst);
      printf
        ("\tedge: { sourcename: \"%s\" targetname: \"%s\" }\n",
         (char *) &src, (char *) &dst);
    }
  else
    {
      /* deferred -- vcg needs all node data in one line */
    }
  return GNUNET_OK;
}


static void *
process (void *cls)
{
  static GNUNET_PeerIdentity null_peer;
  GNUNET_PeerIdentity *current;
  struct GNUNET_ClientServerConnection *sock = cls;
  GNUNET_TRACEKIT_ReportCallback report;
  GNUNET_EncName enc;
  unsigned int i;
  unsigned int j;
  int is_source;
  int is_first;

  report = NULL;
  switch (format)
    {
    case 0:
      report = &human_readable;
      break;
    case 1:
      printf ("digraph G {\n");
      report = &dot_format;
      break;
    case 2:
      report = &vcg_format;
      printf ("graph: {\n");
      break;
    default:
      GNUNET_GE_BREAK (NULL, 0);
      return NULL;
    }
  GNUNET_TRACEKIT_run (sock, depth, priority, report, NULL);
  /* final processing loop */
  for (i = 0; i < count * 2; i++)
    {
      if (0 == i % 2)
        current = &seen[i / 2].src;
      else
        current = &seen[i / 2].dst;
      if (0 == memcmp (current, &null_peer, sizeof (GNUNET_PeerIdentity)))
        continue;
      is_first = GNUNET_YES;
      for (j = 0; j < count * 2; j++)
        if (0 == memcmp (current,
                         (0 == i % 2) ? &seen[i / 2].src : &seen[i / 2].dst,
                         sizeof (GNUNET_PeerIdentity)))
          {
            is_first = GNUNET_NO;
            break;
          }
      if (is_first != GNUNET_YES)
        continue;               /* only each peer once */
      is_source = GNUNET_NO;
      for (j = 0; j < count; j++)
        {
          if (0 == memcmp (current,
                           &seen[i].src, sizeof (GNUNET_PeerIdentity)))
            {
              is_source = GNUNET_YES;
              break;
            }
        }
      switch (format)
        {
        case 0:
          break;
        case 1:
          if (is_source == GNUNET_NO)
            {
              printf ("  \"%.*s\" [style=filled,color=\".7 .3 1.0\"];\n",
                      4, (char *) &enc);
            }
          break;
        case 2:
          if (is_source == GNUNET_NO)
            {
              printf
                ("\tnode: { title: \"%s\" label: \"%.*s\" shape: \"ellipse\" }\n",
                 (char *) &enc, 4, (char *) &enc);
            }
          else
            {
              printf ("\tnode: { title: \"%s\" label: \"%.*s\" }\n",
                      (char *) &enc, 4, (char *) &enc);
            }
          break;
        }
    }
  /* close syntax */
  switch (format)
    {
    case 0:
      break;
    case 1:
      printf ("}\n");
      break;
    case 2:
      printf ("}\n");
      break;
    }
  return NULL;
}

/**
 * All gnunet-tracekit command line options
 */
static struct GNUNET_CommandLineOption gnunettracekitOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'D', "depth", "DEPTH",
   gettext_noop ("probe network to the given DEPTH"), 1,
   &GNUNET_getopt_configure_set_uint, &depth},
  {'F', "format", "FORMAT",
   gettext_noop
   ("specify output format; 0 for human readable output, 1 for dot, 2 for vcg"),
   1,
   &GNUNET_getopt_configure_set_uint, &format},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Start GNUnet transport benchmarking tool.")), /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'P', "priority", "PRIORITY",
   gettext_noop ("use PRIORITY for the priority of the trace request"), 1,
   &GNUNET_getopt_configure_set_uint, &priority,},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  {'W', "wait", "DELAY",
   gettext_noop ("wait DELAY seconds for replies"), 1,
   &GNUNET_getopt_configure_set_uint, &delay},
  GNUNET_COMMAND_LINE_OPTION_END,
};

static void
run_shutdown (void *unused)
{
  GNUNET_shutdown_initiate ();
}

/**
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunet-tracekit: 0: ok, -1: error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_ClientServerConnection *sock;
  struct GNUNET_ThreadHandle *myThread;
  struct GNUNET_CronManager *cron;
  void *unused;

  if (-1 == GNUNET_init (argc,
                         argv,
                         "gnunet-tracekit",
                         &cfgFilename, gnunettracekitOptions, &ectx, &cfg))
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  if (format > 2)
    {
      printf (_("Format specification invalid. "
                "Use 0 for user-readable, 1 for dot, 2 for vcg.\n"));
      return -1;
    }

  sock = GNUNET_client_connection_create (ectx, cfg);
  if (sock == NULL)
    {
      fprintf (stderr, _("Error establishing connection with gnunetd.\n"));
      GNUNET_fini (ectx, cfg);
      return 1;
    }
  myThread = GNUNET_thread_create (&process, sock, 128 * 1024);
  if (myThread == NULL)
    GNUNET_GE_DIE_STRERROR (ectx,
                            GNUNET_GE_FATAL | GNUNET_GE_IMMEDIATE |
                            GNUNET_GE_ADMIN, "pthread_create");
  cron = GNUNET_cron_create (ectx);
  GNUNET_cron_start (cron);
  GNUNET_cron_add_job (cron, &run_shutdown, GNUNET_CRON_SECONDS * delay,
                       0, NULL);
  GNUNET_shutdown_wait_for ();
  GNUNET_client_connection_close_forever (sock);
  GNUNET_thread_join (myThread, &unused);
  GNUNET_client_connection_destroy (sock);
  GNUNET_cron_stop (cron);
  GNUNET_cron_destroy (cron);
  GNUNET_array_grow (seen, count, 0);
  GNUNET_fini (ectx, cfg);
  return 0;
}

/* end of gnunet-tracekit.c */
